tests/ui/run.ts

```diff
@@ -3,59 +3,134 @@
-    await t.expect(firstEmoji.tagName).eql('img');
-    await t.expect(firstEmoji.attributes.title).eql(':woman_zombie:');
-    await t.expect(firstEmoji.classNames).eql(['e1']);
-
-    const secondEmoji = await Selector('div.compose img').nth(1)();
-    await t.expect(secondEmoji.tagName).eql('img');
-    await t.expect(secondEmoji.attributes.title).eql(':beers:');
-    await t.expect(secondEmoji.classNames).eql(['e1']);
-});
+    await driver.findElement(By.className('e1-food')).click();
+    await driver.findElement(By.css('.e1._1f37b')).click();
+
+    // Validate emoji
+    const emoji = await driver.findElement(composeArea).findElements(By.xpath('*'));
+    expect(emoji.length).to.equal(2);
+    expect(await emoji[0].getAttribute('title')).to.equal(':woman_zombie:');
+    expect(await emoji[1].getAttribute('title')).to.equal(':beers:');
+
+    // Validate text
+    const html = await driver.findElement(composeArea).getAttribute('innerHTML');
+    expect(/>hi<img/.test(html)).to.be.true;
+}
+
+/**
+ * Insert a newline using shift-enter.
+ */
+async function insertNewline(driver: WebDriver) {
+    // Insert text
+    await driver.findElement(composeArea).click();
+    await driver.findElement(composeArea).sendKeys('hello');
+    await driver.findElement(composeArea).sendKeys(Key.SHIFT, Key.ENTER);
+    await driver.findElement(composeArea).sendKeys('threema');
+    await driver.findElement(composeArea).sendKeys(Key.SHIFT, Key.ENTER);
+    await driver.findElement(composeArea).sendKeys('web');
+
+    const script = `
+        ${extractText.toString()}
+        const element = document.querySelector("div.compose");
+        return extractText(element);
+    `;
+    const text = await driver.executeScript<string>(script);
+    expect(text).to.equal('hello\nthreema\nweb');
+}
+
+// Register tests here
+const TESTS: Array<[string, Testfunc]> = [
+    ['Show and hide emoji selector', showEmojiSelector],
+    ['Insert emoji and text', insertEmoji],
+    ['Insert three lines of text', insertNewline],
+];
+
+// Test runner
+const TEST_URL = 'http://localhost:7777/tests/ui/compose_area.html';
+(async function() {
+    const driver: WebDriver = await new Builder().forBrowser(browser).build();
+    let i = 0;
+    let success = 0;
+    let failed = 0;
+    console.info('\n====== THREEMA WEB UI TESTS ======\n');
+    try {
+        for (const [name, testfunc] of TESTS) {
+            console.info(TermColor.blue(`» ${i + 1}: Running test: ${name}`));
+            await driver.get(TEST_URL);
+            await testfunc(driver);
+            success++;
+            i++;
+        }
+    } catch (e) {
+        console.error(TermColor.red(`\nTest failed:`));
+        console.error(e);
+        failed++;
+    } finally {
+        await driver.quit();
+    }
+    const colorFunc = failed > 0 ? TermColor.red : TermColor.green;
+    console.info(colorFunc(`\nSummary: ${i} tests run, ${success} succeeded, ${failed} failed`));
```

Comment:
I think I was confused about this debug entry... `i` is logged but not the total amount of tests.

Reply:
Yes, as the text says, it prints how many tests were run, not how many there are. If there are no failures, then all tests will be run.

Reply:
The point being: The code looks like it supports multiple failures while in reality only supports one.